### PR TITLE
[WIP] Add a Bde::pipe method

### PR DESCRIPTION
**Pull request by strk**
The method is still untested, and relies on availability
of /dev/stdout

### CODE REVIEW

File: lib/LINZ/Bde.pm [hunk 1]
```diff
@@ -477,6 +419,137 @@ sub copy
...
+        $log
+    );
+    unlink($cfgtmp) if $cfgtmp;
+
+    my $result = $self->resultFromLog($log);
+    unlink($log) if $opts->{log_file} || ! $opts->{keep_log};
+    return $result;
+}
+
+sub pipe
```
Comment (palmerj) [inline lib/LINZ/Bde.pm]: New function seems good. But before you merge I suggest refactoring the code between the copy and pipe methods as they almost have the same code.

**Comment by strk** [comment]
In an ideal world I'd re-implement "copy" by calling "pipe",
but the "pipe" method only works when "/dev/stdout" exists,
which means not on windows, I guess. Is that acceptable ?

**Comment by strk** [comment]
With 7e4338e I put the options building code into its own function.

**Comment by strk** [comment]
It looks like the method in its current form is broken because it removes the configuration file *before* bde_copy tries to open it (which happens on first read from the filehandle) - See https://github.com/linz/linz_bde_uploader/pull/138#issuecomment-338705123

**Comment by palmerj** [comment]
> In an ideal world I'd re-implement "copy" by calling "pipe",
but the "pipe" method only works when "/dev/stdout" exists,
which means not on windows, I guess. Is that acceptable ?

Yes I think so. Please with @ccrook as this package might need be used on windows for geodetic/tm1 reporting. In the case windows is still requires then does FIFO also work on Perl win32? I see windows has named pipes.

**Comment by palmerj** [comment]
> It looks like the method in its current form is broken because it removes the configuration file before bde_copy tries to open it (which happens on first read from the filehandle) - See linz/linz_bde_uploader#138 (comment)

Iam assuming this is also the same situation wheb using FIFO

**Comment by palmerj** [comment]
> With 7e4338e I put the options building code into its own function.

Thanks @strk

**Comment by strk** [comment]
> Iam assuming this is also the same situation wheb using FIFO

Yes, this has to be fixed in any case.

**Comment by strk** [comment]
@ccrook would you want the new ::pipe method to bail out if `/dev/stdout` does not exist ?
Any chance you can test this PR on windows ?

**Comment by strk** [comment]
With https://github.com/linz/linz-bde-perl/pull/24/commits/15c10224d7b9465fcc56c336a287ff6a7c77509a I've had Bde::pipe die if `/dev/stdout` is not a named pipe. I guess  the method could instead still work by writing a temporary file instead, is that desired (it would probably simplify usage...)

**Comment by ccrook** [comment]
@strk @palmerj Just seen by chance @ mentions of myself in this issue.  Not sure why I've not seen them in my email - will check if I have a misconfigured mail rule.  In response to question on Windows - I think that this can be ignored for this issue.  The geodetic usage replication on Windows is built on an old version of BDE and is shortly to be decommissioned in favour of linux/postgresql platform.  As for TM1 reporting that is also using old code AFAIK, though I'm not so familiar with it.  It is really a question for Noel Tapp but I suspect we don't need to worry about this code for TM1.

**Comment by ccrook** [comment]
@strk - also I don't have a Windows test platform at the moment :-(

**Comment by strk** [comment]
Alright then I'll merge this with the error printed upon usage when `/dev/stdout` is not supported. Can eventually be fixed if anyone complains.

**Comment by strk** [comment]
Squash merged as eba7019acb4a18702ee9b5beffe56cb773640769

**Pull request closed without merging**